arch/risc-v/src/common/espressif/esp_qencoder.c

```diff
@@ -0,0 +1,593 @@
+
+  /* Configure GPIO pins as Input with Pull-Up enabled */
+
+  esp_configgpio(priv->config->ch0_gpio, INPUT_FUNCTION_3 | PULLUP);
+  esp_configgpio(priv->config->ch1_gpio, INPUT_FUNCTION_3 | PULLUP);
+
+  /* Connect Channel A (ch0_gpio) and Channel B (ch1_gpio) crossed for X4 */
+
+  esp_gpio_matrix_in(priv->config->ch0_gpio,
+                     priv->config->ch0_pulse_sig, 0);
+  esp_gpio_matrix_in(priv->config->ch1_gpio,
+                     priv->config->ch0_ctrl_sig, 0);
+
+  esp_gpio_matrix_in(priv->config->ch1_gpio,
+                     priv->config->ch1_pulse_sig, 0);
+  esp_gpio_matrix_in(priv->config->ch0_gpio,
+                     priv->config->ch1_ctrl_sig, 0);
+
+  /* Clear the Reset bit to enable the Pulse Counter */
+
+  pcnt_ll_clear_count(priv->dev, priv->config->pcntid);
+  pcnt_ll_start_count(priv->dev, priv->config->pcntid);
+
+  spin_unlock_irqrestore(&priv->lock, flags);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_shutdown
+ *
+ * Description:
+ *   This method is called when the driver is closed.  The lower half driver
+ *   should stop data collection, free any resources, disable timer hardware,
+ *   and put the system into the lowest possible power usage state
+ *
+ ****************************************************************************/
+
+static int esp_shutdown(struct qe_lowerhalf_s *lower)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
+
+  /* Disable PCNT clock */
+
+  periph_module_disable(PERIPH_PCNT_MODULE);
+
+  /* Make sure initial position is 0 */
+
+  priv->position = 0;
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_position
+ *
+ * Description:
+ *   Return the current position measurement.
+ *
+ ****************************************************************************/
+
+static int esp_position(struct qe_lowerhalf_s *lower, int32_t *pos)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
+  irqstate_t flags;
+  int32_t position;
+  int16_t count;
+
+  DEBUGASSERT(lower && priv->inuse);
+
+  flags = spin_lock_irqsave(&priv->lock);
+
+  position = priv->position;
+  count = (int16_t)(pcnt_ll_get_count(priv->dev, priv->config->pcntid) & 0xffff);
+
+  /* Return the position measurement */
+
+  *pos = position + count;
+
+  spin_unlock_irqrestore(&priv->lock, flags);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_setposmax
+ *
+ * Description:
+ *   Set the maximum encoder position.
+ *
+ ****************************************************************************/
+
+static int esp_setposmax(struct qe_lowerhalf_s *lower, uint32_t pos)
+{
+  return -ENOTTY;
+}
+
+/****************************************************************************
+ * Name: esp_reset
```

Comment:
Missing docstring

Reply:
Could you, please, tell me, how it should look?

Reply:
Sure. You are missing a description of the input parameters and returned value. See this example from `esp_mcpwm.c`

```
/****************************************************************************
 * Name: esp_motor_fault_set
 *
 * Description:
 *   Sets the fault state for the motor.
 *
 * Input Parameters:
 *   lower - Pointer to the motor channel lower-half data structure.
 *   fault - Fault value. Must be one of motor_fault_e enum.
 *
 * Returned Value:
 *   OK on success, otherwise a negated errno value is returned on
 *   any failure.
 *
 ****************************************************************************/
```

Reply:
fixed